include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,275 @@
+/// behavior of a function) increases the major version number. A
+/// cache implementation should check that the API version used to compile
+/// the cache is compatible with the API version of the Triton server
+/// that it is running in. This is typically done by code similar to
+/// the following which makes sure that the major versions are equal
+/// and that the minor version of Triton is >= the minor version used
+/// to build the cache.
+///
+///   uint32_t api_version_major, api_version_minor;
+///   TRITONCACHE_ApiVersion(&api_version_major, &api_version_minor);
+///   if ((api_version_major != TRITONCACHE_API_VERSION_MAJOR) ||
+///       (api_version_minor < TRITONCACHE_API_VERSION_MINOR)) {
+///     return TRITONSERVER_ErrorNew(
+///       TRITONSERVER_ERROR_UNSUPPORTED,
+///       "triton cache API version does not support this cache");
+///   }
+///
+#define TRITONCACHE_API_VERSION_MAJOR 0
+#define TRITONCACHE_API_VERSION_MINOR 1
+
+/// Get the TRITONCACHE API version supported by Triton. This
+/// value can be compared against the
+/// TRITONCACHE_API_VERSION_MAJOR and
+/// TRITONCACHE_API_VERSION_MINOR used to build the cache to
+/// ensure that Triton is compatible with the cache.
+///
+/// \param major Returns the TRITONCACHE API major version supported
+/// by Triton.
+/// \param minor Returns the TRITONCACHE API minor version supported
+/// by Triton.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_ApiVersion(
+    uint32_t* major, uint32_t* minor);
+
+///
+/// CacheEntry Lifetime Management
+///
+
+/// Create a new cache entry object. The caller takes ownership of the
+/// TRITONCACHE_CacheEntry object and must call
+/// TRITONCACHE_CacheEntryDelete to release the object.
+///
+/// \param entry Returns the new cache entry object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryNew(
+    TRITONCACHE_CacheEntry** entry);
+
+/// Delete a cache entry object.
+///
+/// \param entry The cache entry object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryDelete(
+    TRITONCACHE_CacheEntry* entry);
+
+///
+/// CacheEntry Field Management
+///
+
+// TODO: Proper API descriptions, lifetimes, copies, etc.
+
+// Get number of items in entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemCount(
+    TRITONCACHE_CacheEntry* entry, size_t* count);
+
+// Adds item to entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryAddItem(
+    TRITONCACHE_CacheEntry* entry, TRITONCACHE_CacheEntryItem* item);
+
+// Gets item at index from entry where 0 <= index < count and
+// 'count' is the value returned by TRITONCACHE_CacheEntryItemCount
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryGetItem(
```

Comment:
To summarize, a CacheEntry has many CacheEntryItems which can have many CacheEntryBuffers? Why do we need two level hierarchy? Can we flatten the `CacheEntryBuffers` into `CacheEntryItems`?

Reply:
Similar discussion here: https://github.com/triton-inference-server/core/pull/142#discussion_r1043883249